pyarn/lockfile.py

```diff
@@ -63,19 +80,38 @@ def from_dict(cls, raw_name, data):
             alias = name
             name, _version = _must_match(name_at_version, _remove_prefix(_version, "npm:")).groups()
 
-        if _version and _version.startswith("file:"):
-            path = _remove_prefix(_version, "file:")
+        if _version:
+            path = cls.get_path_from_version_specifier(_version)
+
+        # Ensure the resolved version key exists to appease mypy
+        version = data.get("version")
+        if not version:
+            raise ValueError("Package version was not provided")
 
         return cls(
             name=name,
-            version=data.get("version"),
+            version=version,
             url=data.get("resolved"),
             checksum=data.get("integrity"),
-            relpath=path,
+            path=path,
             dependencies=data.get("dependencies", {}),
             alias=alias,
         )
 
+    @staticmethod
+    def get_path_from_version_specifier(version: str) -> Optional[str]:
+        "Return the path from a package.json file dependency version specifier."
+        version_path = Path(version)
+
+        if version.startswith("file:"):
+            return _remove_prefix(version, "file:")
+        elif version.startswith("link:"):
+            return _remove_prefix(version, "link:")
+        elif version_path.is_absolute() or version.startswith(("./", "../")):
+            return str(version_path)
+        else:
+            return None
```

Comment:
A minor, non-blocking suggestion: could you please add a short comment or a link to a doc about when this branch could be reached?